### DISCUSSION TITLE
Documentation for creating custom Docker and qemu images?

### DISCUSSION THREAD

**Discussion by jaskij** [cross-rs/cross#700] [Q&A]
For testing a project, I need create custom qemu images. What I gather is that for this to work with cross I would need to *also* have a custom Docker image.

What are the required packages, in both images, to make this work?

P.S.
The project needs custom qemu images because it needs to be tested on specific kernel versions.

**Comment by Emilgardis** [comment]
I'm unsure what you mean with `both images`, but it should just suffice with installing your qemu version on the image.

```dockerfile
# :latest is not available right now, use main or edge
FROM ghcr.io/cross-rs/aarch64-unknown-linux-gnu:main

COPY ./my-qemu-pkg.pkg my-qemu-pkg.pkg 
RUN pseudo-install my-qemu-pkg.pkg 
```

simpler solution may be:

fork the project, modify the current [qemu script](https://github.com/cross-rs/cross/blob/main/docker/qemu.sh) and run [`build-docker-image.sh x86_64-unknown-linux-gnu`](https://github.com/cross-rs/cross/blob/main/build-docker-image.sh) to get your own image. Make sure to rename the package name [here](https://github.com/cross-rs/cross/blob/7599e1a513823a02ccdaf510a5f63c206c83f31a/build-docker-image.sh#L31)

**Reply by jaskij** [reply]
`both images` meaning both qemu and docker images. I probably muddled the waters by including Docker in my question. What I primarily care about is having my own qemu images which can be used with cross, and having cross somehow use them.

Also, why do you metion `.pkg`? By image I want the image to run under qemu, not a custom qemu **version.**

Also, while I find your scripts somewhat hard to read, from what I gathered you use the distribution-provided kernel? That is the opposite of what I want.

**Reply by Emilgardis** [reply]
aha! Right, I think I understand now. There are no required dependencies really, all that is needed is that the image is somehow able to run x86_64-unknown-linux-gnu `cargo` and `rustc`. If you have an image with the kernel and qemu you want, you should be able to just use them.

https://github.com/cross-rs/cross#configuration

```
[target.{target}]
image = "my/qemuimage:tag"
```

make sure the image sets the [appropriate env flags](https://doc.rust-lang.org/cargo/reference/config.html#targettriplerunner) to tell cargo how to run the executable if you're using `cross run`

We do a lot of setup in our images to ensure some things work as expected but it's all really just fluff. cross is essentially an easy way to mount and run a rust toolchain target into an image and invoke cargo on it, see how with `cross build -vv` for example
